packages/client-config/src/client-config-contributor/custom_client_config_contributor.ts

```diff
@@ -0,0 +1,68 @@
+import { ClientConfigContributor } from './client_config_contributor.js';
+import { UnifiedBackendOutput } from '@aws-amplify/backend-output-schemas';
+import { CustomClientConfig } from '../client-config-types/client_config.js';
+
+// TODO: these types are copied, should be moved to platform-types perhaps.
+export type ClientConfigDestination = {
+  clientConfigFormat: string;
+  path: Array<string>;
+}
+
+export type CustomOutputOptions = {
+  clientConfigDestinations?: Array<ClientConfigDestination>
+}
+
+export class CustomClientConfigContributor implements ClientConfigContributor {
+  contribute = async ({
+    ['AWS::Amplify::Custom']: customOutputs,
+  }: UnifiedBackendOutput): Promise<
+    CustomClientConfig | Record<string, never>
+  > => {
+    if (
+      customOutputs === undefined ||
+      Object.keys(customOutputs.payload).length === 0
+    ) {
+      console.log('custom outputs undefined');
+      return {};
+    }
+
+    console.log(customOutputs);
+
+    const result: CustomClientConfig & Record<string, unknown> = { custom: {} };
+
+    Object.entries(customOutputs.payload).forEach(([key, value]) => {
+      const customOutputData = JSON.parse(value);
+      // TODO how do we do custom namespacing ??
+      result.custom[key.replace('amplifycustom', '')] = customOutputData.value;
+
+      const customOutputOptions: CustomOutputOptions | undefined = customOutputData.options;
+
+      if (customOutputOptions){
+        // TODO should this be in addition to placing value in custom.X section or should these be XOR ?
+        customOutputOptions.clientConfigDestinations?.forEach(clientConfigDestination => {
+          let currentRef: Record<string, unknown> = result;
+          let currentArrayRef: Array<unknown> | undefined = undefined;
+          for (let i = 0 ; i < clientConfigDestination.path.length - 1; i++) {
+            const pathSegment = clientConfigDestination.path[i];
+            if (!currentRef[pathSegment]) {
+              currentRef[pathSegment] = {}
+            }
+            currentRef = currentRef[pathSegment] as Record<string, unknown>;
+          }
+          const lastPathSegment = clientConfigDestination.path[clientConfigDestination.path.length - 1];
+          // TODO this is a hack
+          // figure out how do we deal with arrays.
+          if (lastPathSegment.endsWith('[]')) {
+            currentRef[lastPathSegment.replace('[]', '')] = [customOutputData.value];
+          }else {
+            currentRef[lastPathSegment] = customOutputData.value;
+          }
+        });
+      }
+    });
+
+    console.log(result);
+
+    return result;
+  };
```

Comment:
This file is interesting.

The idea is that:
1. Custom outputs are put into `custom.X` path in client config by default
2. If directive was sent from `setCustomOutput` api these values can be placed into specific locations (i.e. to handle Geo use case where we need to match schema expected by frontend libs that exist today).

The actual mechanism to target specific config keys is TBD and a bit hacky as of now.

Reply:
Only thing I don't see here is how existing config values can be overwritten. Seems like we might need to have this custom contributor run last in the "contributor pipeline" so that it can apply customer specified overrides to existing values. Or maybe there's some other mechanism we're going to use for that?

Reply:
Yep. It's a position of the contributor in the pipeline (being last) that makes it work in "override scenario".

And that can remain as a solution if we decide that "path" should be JS path (or whatever we think the unified format is at given point of time).

This can get complicate if we allow "path per config type", in that case we'd need to build some mechanism to override js->other type result. (but maybe we'll never get there, so I not implemented that yet).

Reply:
https://github.com/aws-amplify/amplify-backend/pull/946#discussion_r1467159112